README.md
# MMM-json

A Module for [MagicMirror](https://github.com/MichMich/MagicMirror) designed to
display parts or the whole JSON response from an api.

## Preview

![alt text](https://github.com/DanielHabenicht/MMM-json/raw/main/sample.png "Example")

![alt text](https://user-images.githubusercontent.com/13590797/148958016-69f48869-8bec-4d90-b221-10025de54503.png "Advanced Example")



## Installation

[Jump to the wiki](https://danielhabenicht.github.io/MMM-json/) for installation and configuration. 


changelog.md
# [2.3.0](https://github.com/DanielHabenicht/MMM-json/compare/v2.2.0...v2.3.0) (2023-01-14)


### Features

* add option to add icon to any property ([f3d3257](https://github.com/DanielHabenicht/MMM-json/commit/f3d325746fdfb7550780bcda12a5ed195c1e415a))

# [2.2.0](https://github.com/DanielHabenicht/MMM-json/compare/v2.1.0...v2.2.0) (2022-01-09)


### Features

* add multi value support ([e5af63a](https://github.com/DanielHabenicht/MMM-json/commit/e5af63add94112f4fc850dbe2d4a435b4df1ee54))

# [2.1.0](https://github.com/DanielHabenicht/MMM-json/compare/v2.0.0...v2.1.0) (2022-01-09)


### Bug Fixes

* **test:** add more tests ([e300842](https://github.com/DanielHabenicht/MMM-json/commit/e3008427b4aec0dcb8ffb853f5bafb018a281bca))


### Features

* add style rules ([#10](https://github.com/DanielHabenicht/MMM-json/issues/10)) ([949ff92](https://github.com/DanielHabenicht/MMM-json/commit/949ff926b1a9dc2875c149c7f24e03fd9b966c98))

# [2.0.0](https://github.com/DanielHabenicht/MMM-json/compare/v1.1.4...v2.0.0) (2022-01-09)


### Bug Fixes

* Rename Notifications ([869e182](https://github.com/DanielHabenicht/MMM-json/commit/869e1828f886e0a0e87f54eb726d84a4a4b62ba8))
* use node-fetch instead of request ([75cd82f](https://github.com/DanielHabenicht/MMM-json/commit/75cd82f314dfc6601e51ad81ae3af8aee949f63e)), closes [#7](https://github.com/DanielHabenicht/MMM-json/issues/7)


### BREAKING CHANGES

* use `fetchOptions` instead of `request` in you config
* rename notifications

to better resemble that they could also be POST or other requests.

MMM_JSON_GET_REQUEST > MMM_JSON_REQUEST
MMM_JSON_GET_RESPONSE > MMM_JSON_RESPONSE

## [1.1.4](https://github.com/DanielHabenicht/MMM-json/compare/v1.1.3...v1.1.4) (2022-01-09)


### Bug Fixes

* **release:** fix releaserc ([1e79db4](https://github.com/DanielHabenicht/MMM-json/commit/1e79db40581f5cee33f1902443e62419a948510e))

# 1.0.0 (2021-03-03)


### Bug Fixes

* **header:** show builtin header ([b23d011](https://github.com/DanielHabenicht/MMM-json/commit/b23d011914edfcd43d472f8377b79b38283c353c)), closes [#1](https://github.com/DanielHabenicht/MMM-json/issues/1)


MMM-json.spec.js
/* eslint no-multi-spaces: 0 */
const expect = require("chai").expect;
var jsdom = require("mocha-jsdom");
let modulename = "MMM-json";
let identifier = "test";

describe("Functions module MMM-json", function () {
  before(function () {
    Module = {};
    Module.definitions = {};
    Module.register = function (name, moduleDefinition) {
      Module.definitions[name] = moduleDefinition;
    };
    require("./MMM-json");
    Module.definitions[modulename].updateDom = () => {
      // noop
    };
    Module.definitions[modulename].config = {};
    Module.definitions[modulename].identifier = identifier;
  });

  describe("getHeader", function () {
    describe("with header icon", function () {
      before(function () {
        Module.definitions[modulename].config.headerIcon = "icon";
      });

      it(`should be empty`, function () {
        expect(Module.definitions[modulename].getHeader()).to.equal("");
      });
    });

    describe("without header icon, but header", function () {
      before(function () {
        delete Module.definitions[modulename].config.headerIcon;
        Module.definitions[modulename].data = {};
        Module.definitions[modulename].data.header = "I am a Header";
      });

      it(`should contain header`, function () {
        expect(Module.definitions[modulename].getHeader()).to.equal(
          "I am a Header"
        );
      });
    });
    describe("without header icon and no header", function () {
      before(function () {
        delete Module.definitions[modulename].config.headerIcon;
        Module.definitions[modulename].data = {};
        delete Module.definitions[modulename].data.header;
      });

      it(`should contain empty header`, function () {
        expect(Module.definitions[modulename].getHeader()).to.equal("");
      });
    });
  });

  describe("dom", function () {
    jsdom({
      url: "https://localhost"
    });
    describe("config", function () {
      it(`missing - null`, function () {
        delete Module.definitions[modulename].config.url;
        expect(Module.definitions[modulename].getDom().outerHTML).to.equal(
          "<div>Missing configuration.</div>"
        );
      });
      it(`missing - empty`, function () {
        Module.definitions[modulename].config.url = "";
        expect(Module.definitions[modulename].getDom().outerHTML).to.equal(
          "<div>Missing configuration.</div>"
        );
      });
      it(`loading`, function () {
        Module.definitions[modulename].config.url = "https://localhost";
        expect(Module.definitions[modulename].getDom().outerHTML).to.equal(
          "<div>Loading...</div>"
        );
      });
    });
    describe("data", function () {
      before(function () {
        Module.definitions[modulename].config.url = "https://localhost";
      });
      it(`normal`, function () {
        Module.definitions[modulename].socketNotificationReceived(
          "MMM_JSON_RESPONSE",
          {
            identifier: identifier,
            data: [{ title: "test-title", value: ["test-value"] }],
            error: false
          }
        );

        expect(Module.definitions[modulename].getDom().outerHTML).to.equal(
          '<div><table><tr><td class="small regular bright">test-title:</td><td class="small light bright">test-value</td></tr></table></div>'
        );
      });
      it(`basic - prefix`, function () {
        Module.definitions[modulename].socketNotificationReceived(
          "MMM_JSON_RESPONSE",
          {
            identifier: identifier,
            data: [
              { title: "test-title", value: ["test-value"], prefix: "PREFIX" }
            ],
            error: false
          }
        );
        expect(Module.definitions[modulename].getDom().outerHTML).to.equal(
          '<div><table><tr><td class="small regular bright">test-title:</td><td class="small light bright">PREFIX test-value</td></tr></table></div>'
        );
      });
      it(`basic - suffix`, function () {
        Module.definitions[modulename].socketNotificationReceived(
          "MMM_JSON_RESPONSE",
          {
            identifier: identifier,
            data: [
              { title: "test-title", value: ["test-value"], suffix: "SUFFIX" }
            ],
            error: false
          }
        );
        expect(Module.definitions[modulename].getDom().outerHTML).to.equal(
          '<div><table><tr><td class="small regular bright">test-title:</td><td class="small light bright">test-value SUFFIX</td></tr></table></div>'
        );
      });

      it(`basic - both`, function () {
        Module.definitions[modulename].socketNotificationReceived(
          "MMM_JSON_RESPONSE",
          {
            identifier: identifier,
            data: [
              {
                title: "test-title",
                value: ["test-value"],
                prefix: "PREFIX",
                suffix: "SUFFIX"
              }
            ],
            error: false
          }
        );
        expect(Module.definitions[modulename].getDom().outerHTML).to.equal(
          '<div><table><tr><td class="small regular bright">test-title:</td><td class="small light bright">PREFIX test-value SUFFIX</td></tr></table></div>'
        );
      });

      it(`basic - all`, function () {
        Module.definitions[modulename].socketNotificationReceived(
          "MMM_JSON_RESPONSE",
          {
            identifier: identifier,
            data: [
              {
                title: "title1",
                value: ["value1"],
                prefix: "PREFIX",
                suffix: "SUFFIX"
              },
              {
                title: "title2",
                value: ["value2"],
                prefix: ["PREFIX"],
                suffix: ["SUFFIX"]
              },
              {
                title: "title3",
                value: ["value3a", "value3b"],
                prefix: ["PREFIXa", "PREFIXb"],
                suffix: ["SUFFIXa", "SUFFIXb"]
              },
              {
                title: "title4",
                value: ["value4a", "value4b"],
                prefix: "PREFIX",
                suffix: "SUFFIX"
              }
            ],
            error: false
          }
        );
        expect(Module.definitions[modulename].getDom().outerHTML).to.equal(
          '<div><table><tr><td class="small regular bright">title1:</td><td class="small light bright">PREFIX value1 SUFFIX</td></tr><tr><td class="small regular bright">title2:</td><td class="small light bright">PREFIX value2 SUFFIX</td></tr><tr><td class="small regular bright">title3:</td><td class="small light bright">PREFIXa value3a SUFFIXa</td><td class="small light bright">PREFIXb value3b SUFFIXb</td></tr><tr><td class="small regular bright">title4:</td><td class="small light bright">PREFIX value4a SUFFIX</td><td class="small light bright">PREFIX value4b SUFFIX</td></tr></table></div>'
        );
      });
    });
  });
});


testing/class.js
/* global Class, xyz */

/* Simple JavaScript Inheritance
 * By John Resig https://johnresig.com/
 *
 * Inspired by base2 and Prototype
 *
 * MIT Licensed.
 */
(function () {
  let initializing = false;
  const fnTest = /xyz/.test(function () {
    xyz;
  })
    ? /\b_super\b/
    : /.*/;

  // The base Class implementation (does nothing)
  this.Class = function () {};

  // Create a new Class that inherits from this class
  Class.extend = function (prop) {
    let _super = this.prototype;

    // Instantiate a base class (but only create the instance,
    // don't run the init constructor)
    initializing = true;
    const prototype = new this();
    initializing = false;

    // Make a copy of all prototype properties, to prevent reference issues.
    for (const p in prototype) {
      prototype[p] = cloneObject(prototype[p]);
    }

    // Copy the properties over onto the new prototype
    for (const name in prop) {
      // Check if we're overwriting an existing function
      prototype[name] =
        typeof prop[name] === "function" &&
        typeof _super[name] === "function" &&
        fnTest.test(prop[name])
          ? (function (name, fn) {
              return function () {
                const tmp = this._super;

                // Add a new ._super() method that is the same method
                // but on the super-class
                this._super = _super[name];

                // The method only need to be bound temporarily, so we
                // remove it when we're done executing
                const ret = fn.apply(this, arguments);
                this._super = tmp;

                return ret;
              };
            })(name, prop[name])
          : prop[name];
    }

    /**
     * The dummy class constructor
     */
    function Class() {
      // All construction is actually done in the init method
      if (!initializing && this.init) {
        this.init.apply(this, arguments);
      }
    }

    // Populate our constructed prototype object
    Class.prototype = prototype;

    // Enforce the constructor to be what we expect
    Class.prototype.constructor = Class;

    // And make this class extendable
    Class.extend = arguments.callee;

    return Class;
  };
})();

/**
 * Define the clone method for later use. Helper Method.
 *
 * @param {object} obj Object to be cloned
 * @returns {object} the cloned object
 */
function cloneObject(obj) {
  if (obj === null || typeof obj !== "object") {
    return obj;
  }

  const temp = obj.constructor(); // give temp the original obj's constructor
  for (const key in obj) {
    temp[key] = cloneObject(obj[key]);

    if (key === "lockStrings") {
      Log.log(key);
    }
  }

  return temp;
}

/*************** DO NOT EDIT THE LINE BELOW ***************/
if (typeof module !== "undefined") {
  module.exports = Class;
}


testing/logger.js
/* Magic Mirror
 * Log
 *
 * This logger is very simple, but needs to be extended.
 * This system can eventually be used to push the log messages to an external target.
 *
 * By Michael Teeuw https://michaelteeuw.nl
 * MIT Licensed.
 */
(function (root, factory) {
  if (typeof exports === "object") {
    if (process.env.JEST_WORKER_ID === undefined) {
      // add timestamps in front of log messages
      require("console-stamp")(console, {
        pattern: "yyyy-mm-dd HH:MM:ss.l",
        include: ["debug", "log", "info", "warn", "error"]
      });
    }
    // Node, CommonJS-like
    module.exports = factory(root.config);
  } else {
    // Browser globals (root is window)
    root.Log = factory(root.config);
  }
})(this, function (config) {
  let logLevel;
  let enableLog;
  if (typeof exports === "object") {
    // in nodejs and not running with jest
    enableLog = process.env.JEST_WORKER_ID === undefined;
  } else {
    // in browser and not running with jsdom
    enableLog = typeof window === "object" && window.name !== "jsdom";
  }

  if (enableLog) {
    logLevel = {
      debug: Function.prototype.bind.call(console.debug, console),
      log: Function.prototype.bind.call(console.log, console),
      info: Function.prototype.bind.call(console.info, console),
      warn: Function.prototype.bind.call(console.warn, console),
      error: Function.prototype.bind.call(console.error, console),
      group: Function.prototype.bind.call(console.group, console),
      groupCollapsed: Function.prototype.bind.call(
        console.groupCollapsed,
        console
      ),
      groupEnd: Function.prototype.bind.call(console.groupEnd, console),
      time: Function.prototype.bind.call(console.time, console),
      timeEnd: Function.prototype.bind.call(console.timeEnd, console),
      timeStamp: Function.prototype.bind.call(console.timeStamp, console)
    };

    logLevel.setLogLevel = function (newLevel) {
      if (newLevel) {
        Object.keys(logLevel).forEach(function (key, index) {
          if (!newLevel.includes(key.toLocaleUpperCase())) {
            logLevel[key] = function () {};
          }
        });
      }
    };
  } else {
    logLevel = {
      debug: function () {},
      log: function () {},
      info: function () {},
      warn: function () {},
      error: function () {},
      group: function () {},
      groupCollapsed: function () {},
      groupEnd: function () {},
      time: function () {},
      timeEnd: function () {},
      timeStamp: function () {}
    };

    logLevel.setLogLevel = function () {};
  }

  return logLevel;
});


node_helper.js
const fetch = (...args) =>
  import("node-fetch").then(({ default: fetch }) => fetch(...args));
var jp = require("jsonpath");
const { jq } = require("jq.node");
var NodeHelper = require("node_helper");

function asPromise(context, callbackFunction, ...args) {
  return new Promise((resolve, reject) => {
    args.push((err, data) => {
      if (err) {
        reject(err);
      } else {
        resolve(data);
      }
    });
    if (context) {
      callbackFunction.call(context, ...args);
    } else {
      callbackFunction(...args);
    }
  });
}

async function do_jq(filter, data) {
  if (filter) {
    try {
      data = await asPromise(null, jq, JSON.stringify(data), filter, {});
      data = JSON.parse(data);
      console.debug("JQ Converted data: ", data);
    } catch (e) {
      console.error("Error handling jq.node filter '" + filter + "':", e);
    }
  }
  return data;
}

module.exports = NodeHelper.create({
  start: function () {
    console.log("Starting node helper: " + this.name);
  },

  socketNotificationReceived: function (notification, payload) {
    var self = this;
    console.log("Notification: " + notification + " Payload:", payload);

    if (notification === "MMM_JSON_REQUEST") {
      fetch(payload.config.url, payload.config.fetchOptions)
        .then(async (response) => {
          // Success
          console.debug(response);
          if (Math.floor(response.status / 100) === 2) {
            var responseObject;

            jsonData = await do_jq(payload.config.jq, await response.json());
            if (
              payload.config.values == undefined ||
              payload.config.values.length == 0
            ) {
              // Values are not defined return all object properties
              responseObject = {
                identifier: payload.identifier,
                data: Object.keys(jsonData).map((prop) => {
                  return { title: prop, value: [jsonData[prop]] };
                })
              };
            } else {
              // Values are defined, get what the user wants
              responseObject = {
                identifier: payload.identifier,
                data: payload.config.values.map((val) => {
                  return {
                    ...val,
                    value: (Array.isArray(val.query)
                      ? val.query
                      : [val.query]
                    ).map((query) => {
                      return val.numberDevisor != undefined
                        ? (
                            jp.query(jsonData, query)[0] / val.numberDevisor
                          ).toFixed(3)
                        : jp.query(jsonData, query)[0];
                    })
                  };
                })
              };
            }

            self.sendSocketNotification("MMM_JSON_RESPONSE", responseObject);
          }
        })
        .catch((error) => {
          console.error(error);
          // Error
          self.sendSocketNotification("MMM_JSON_RESPONSE", {
            identifier: payload.identifier,
            error: true
          });
          console.error(self.name + " error:", error);
        });
    }
  }
});


MMM-SolarKiwiGrid.js
/*
 * Magic Mirror module for displaying Enphase Solar data
 * By Thomas Krywitsky
 * MIT Licensed
 */

Module.register("MMM-SolarKiwiGrid", {
  // Default module config.
  defaults: {
    url: "https://api.enphaseenergy.com/api/v2/systems/",
    refreshInterval: 1000 * 60 * 5 //5 minutes
  },

  start: function () {
    Log.info("Starting module: " + this.name);
    this.titles = ["Add some values"];
    this.suffixes = ["%"];
    this.results = ["Loading"];
    this.loaded = false;
    this.getSolarData();

    var self = this;
    //Schedule updates
    setInterval(function () {
      self.getSolarData();
      self.updateDom();
    }, this.config.refreshInterval);
  },

  //Import additional CSS Styles
  getStyles: function () {
    return ["solar.css"];
  },

  //Contact node helper for solar data
  getSolarData: function () {
    Log.info("SolarApp: getting data");

    this.sendSocketNotification("GET_SOLAR", {
      config: this.config
    });
  },

  //Handle node helper response
  socketNotificationReceived: function (notification, payload) {
    if (notification === "SOLAR_DATA") {
      this.data = payload;
      this.loaded = true;
      this.updateDom(1000);
    }
  },

  // Override dom generator.
  getDom: function () {
    var wrapper = document.createElement("div");
    if (this.config.url === "") {
      wrapper.innerHTML = "Missing configuration.";
      return wrapper;
    }

    //Display loading while waiting for API response
    if (!this.loaded) {
      wrapper.innerHTML = "Loading...";
      return wrapper;
    }

    var tb = document.createElement("table");

    if (this.config.header) {
      var imgDiv = document.createElement("div");

      var sTitle = document.createElement("p");
      sTitle.innerHTML = this.config.header;
      sTitle.className += "normal";

      if (this.config.headerIcon) {
        var icon = document.createElement("i");
        icon.className = "fas " + this.config.headerIcon;
        sTitle.style = "margin: 0px 0px 0px 15px;";
        imgDiv.appendChild(icon);
      }
      imgDiv.appendChild(sTitle);

      var divider = document.createElement("hr");
      divider.className += " dimmed";
      wrapper.appendChild(imgDiv);
      wrapper.appendChild(divider);
    }

    for (var i = 0; i < this.data.length; i++) {
      var row = document.createElement("tr");

      var titleTr = document.createElement("td");
      var dataTr = document.createElement("td");

      titleTr.innerHTML = this.data[i].title + ":";
      dataTr.innerHTML =
        this.data[i].value +
        " " +
        (this.data[i].suffix ? this.data[i].suffix : "");

      titleTr.className += " small regular bright";
      dataTr.className += " small light bright";

      row.appendChild(titleTr);
      row.appendChild(dataTr);

      tb.appendChild(row);
    }
    wrapper.appendChild(tb);
    return wrapper;
  }
});


MMM-json.js
/*
 * Magic Mirror module for displaying JSON values
 * By Daniel Habenicht
 * MIT Licensed
 */

Module.register("MMM-json", {
  // Default module config.
  defaults: {
    url: "https://jsonplaceholder.typicode.com/users",
    refreshInterval: 1000 * 60 * 5, // 5 minutes
    styleRules: []
  },

  start: function () {
    Log.info("Starting module: " + this.name);
    this.loaded = false;
    this.getData();

    var self = this;
    // Schedule updates
    setInterval(function () {
      self.getData();
      self.updateDom();
    }, this.config.refreshInterval);
  },

  // Import additional CSS Styles
  getStyles: function () {
    return ["mmm-json.css"];
  },

  // Contact node helper for data
  getData: function () {
    Log.info("MMM-json: getting data");

    this.sendSocketNotification("MMM_JSON_REQUEST", {
      config: this.config,
      identifier: this.identifier
    });
  },

  // Handle node helper response
  socketNotificationReceived: function (notification, payload) {
    if (
      notification === "MMM_JSON_RESPONSE" &&
      payload.identifier === this.identifier
    ) {
      if (payload.error === true) {
        console.error(
          "MMM-JSON: An Error occured while fetching your response. Please have a look at the server log."
        );
        this.loaded = false;
      } else {
        this.loaded = true;
        this.response = payload.data;
      }
      this.updateDom(1000);
    }
  },
  // Override the Header generator
  getHeader: function () {
    // If an Icon should be displayed we need our own header
    if (this.config.headerIcon) {
      return "";
    } else {
      return this.data.header || "";
    }
  },

  // Override dom generator.
  getDom: function () {
    var wrapper = document.createElement("div");
    if (this.config.url == null || this.config.url === "") {
      wrapper.innerHTML = "Missing configuration.";
      return wrapper;
    }

    // Display loading while waiting for API response
    if (!this.loaded) {
      wrapper.innerHTML = "Loading...";
      return wrapper;
    }

    var tb = document.createElement("table");

    // Display our own header if we want an icon
    if (this.config.headerIcon) {
      var imgDiv = document.createElement("div");

      var sTitle = document.createElement("p");
      sTitle.innerHTML = this.data.header || "";
      sTitle.className += "normal";

      var icon = document.createElement("i");
      icon.className = "fas " + this.config.headerIcon;
      sTitle.style = "margin: 0px 0px 0px 15px;";
      imgDiv.appendChild(icon);
      imgDiv.appendChild(sTitle);

      var divider = document.createElement("hr");
      divider.className += "dimmed";
      wrapper.appendChild(imgDiv);
      wrapper.appendChild(divider);
    }

    for (var i = 0; i < this.response.length; i++) {
      var row = document.createElement("tr");

      var keyTd = document.createElement("td");
	  keyTd.className = "small light bright";

      if (this.response[i].icon) {
        var icon = document.createElement("i");
        icon.className = this.response[i].icon;
        keyTd.appendChild(icon);
      }

      var titleSpan = document.createElement("span");
      if (this.response[i].title) {
        titleSpan.className = "small regular bright";
        titleSpan.innerHTML =
          (this.response[i].title ? this.response[i].title : "") + ":";
      }
      keyTd.appendChild(titleSpan);
	  row.appendChild(keyTd);

      for (var j = 0; j < this.response[i].value.length; j++) {
        var dataTr = document.createElement("td");

        // Add Prefix
        if (
          this.response[i].prefix != null &&
          (!Array.isArray(this.response[i].prefix) ||
            this.response[i].prefix[j])
        ) {
          dataTr.innerHTML += Array.isArray(this.response[i].prefix)
            ? this.response[i].prefix[j]
            : this.response[i].prefix;
          dataTr.innerHTML += " ";
        }

        // Add Value
        dataTr.innerHTML += this.response[i].value[j];

        // Add Suffix
        if (
          this.response[i].suffix != null &&
          (!Array.isArray(this.response[i].suffix) ||
            this.response[i].suffix[j])
        ) {
          dataTr.innerHTML += " ";
          dataTr.innerHTML += Array.isArray(this.response[i].suffix)
            ? this.response[i].suffix[j]
            : this.response[i].suffix;
        }

        // Add Matching Styles
        if (
          this.config.styleRules != null &&
          this.config.styleRules.length > 0
        ) {
          for (var k = 0; k < this.config.styleRules.length; k++) {
            if (this.config.styleRules[k].match(this.response[i].value[j])) {
              if (this.config.styleRules[k].style != null) {
                dataTr.style = this.config.styleRules[k].style;
              }
              if (this.config.styleRules[k].class != null) {
                dataTr.className += " " + this.config.styleRules[k].class;
              }
            }
          }
        }
        dataTr.className = "small light bright";
        row.appendChild(dataTr);
      }

      tb.appendChild(row);
    }
    wrapper.appendChild(tb);
    return wrapper;
  }
});
